Fix view behavior for AwkwardArrays

grst opened the pull request:
- [x] Closes #1035
- [x] Follow up of #1040 
- [ ] Release note added (or unnecessary)

grst committed:
Use __record__ instead of __list__ to store behavior

grst commented:
I think this does the trick

grst committed:
Fix views of views

flying-sheep reviewed (changes requested):
Shouldn’t this have some tests?

agoose77 commented on anndata/_core/views.py:
```diff
@@ -308,15 +308,19 @@ def as_view_awkarray(array, view_args):
+        #
+        # Allow only Arrays that have no record behavior attached, or with __record__ explicitly
+        # set to None
+        if ak.parameters(array).get("__record__", None) is not None:
```
(Record) names can be set at arbitrary depths. I'd suggest checking recursively using our L2 (public-developer) API:
```suggestion
        if array.layout.purelist_parameter("__record__") is not None:
```

agoose77 replied:
[This setitem](https://github.com/scverse/anndata/blob/b6406179f50ce69d84112a5426d206a203a1181d/anndata/_core/views.py#L323) is responsible for registering the behavior class for lists containing records, rather than 1D arrays of records.

grst committed:
Update anndata/_core/views.py

Co-authored-by: Angus Hollands <goosey15@gmail.com>

agoose77 commented on anndata/_core/views.py:
```diff
@@ -295,8 +295,8 @@ def __copy__(self) -> AwkArray:
-            array = ak.with_parameter(self, _PARAM_NAME, None)
-            array = ak.with_parameter(array, "__list__", None)
+            array = ak.with_parameter(array, _PARAM_NAME, None)
+            array = ak.with_parameter(array, "__record__", None)
```
This should use `with_name`, so that we can traverse into the layout to find the record-node
```suggestion
            array = ak.with_name(array, None)
```

ilan-gold replied:
I imagine that after our dicsussion @grst could not get that working (see https://github.com/scverse/anndata/pull/1040#issuecomment-1647878246)

Is there a reason traversing is better than what we have (assuming @grst could not get the `with_name` working)?  We only ever deal with this at a "top level."

agoose77 commented:
@grst I get the sense that it might be helpful to have a meeting to discuss this. Would you be able to find time for a zoom at some point?

grst commented:
Agreed, I sent an email with a poll to find a timeslot.

grst committed:
Remove all custom code for views

grst committed:
WIP add test cases

grst committed:
Merge remote-tracking branch 'origin/main' into ak-array-views

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

grst committed:
More test cases

grst committed:
Add comment about view behavior

grst commented:
@ivirshup, @ilan-gold, @agoose77, ready for re-review. Feels good to get rid of all that code :boom: 

Currently the tests fail for `uns`, but that seems to be a more general problem, see #1118.

agoose77 commented:
@grst it is possible to modify the original array if you don't perform any row-based slices, i.e `array["x"] = x`, then this would modify `array`.

I can't recall to what extent we touched on this on the call. If the process of pulling out the Awkward Array always returns a shallow copy, this wouldn't be a problem.

grst commented:
Hi @agoose77, 

returning the shallow copy (or slice) works as expected. The remaining problem is that there's no handler implemented for the `uns` slot while subsetting AnnData -- regardless of whether it's an awkward array or a numpy array. I don't think we want to always return a shallow copy on `__getitem__` -- I think there are legitimate use-cases where one wants to modify records.

ivirshup commented:
> The remaining problem is that there's no handler implemented for the uns slot while subsetting AnnData

Is that blocking for this PR, or tangential? I'd like to make a release candidate by the end of the week, so would like to merge soon if possible.

grst commented:
If you don't mind, I can mark the respective tests as `xfail`. I think it's ok as
 * the behavior is not a regression of this PR, it was like that before
 * the problem is tracked in #1118 and
 * I don't think it affects many users

grst committed:
Merge branch 'main' into ak-array-views

ivirshup commented:
> I can mark the respective tests as xfail

Which tests are these?

grst commented:
@ivirshup, these: 
https://github.com/scverse/anndata/pull/1070/files#diff-1534693aa7219116bb36568d007d5f2116414d0c86148cfa15d859b861abafc7R143-R166

grst committed:
Fix akward array view test

The test didn't catch that when modifying an awkward array in a
AnnDataView, the changes did not persist within that view.

The issue was that I had been retreiving the awkward array
from the view once, and then did tests on it. Instead, to properly
test this, I need to retreive the Awkward array from the AnnData view
every time. It turns out doing so gives me a fresh copy of the origina
awkward array and modifications only affect that copy - not the original data.

grst committed:
Fix that awkward arrays in an awyward array view could not be modified.

I solved the issue described in the previous commit by adding
logic to `AlignedViewMixin` that can copy objects of a certain type
(for now only awkward arrays) upon view creation.

grst commented:
Thanks for catching this! I thought my tests cover this, but it turns out I retrieved the awkward array from the view only once and did all tests on that copy. Instead, retreiving `v.obsm["awk"]` multiple times always returns a fresh copy of the sliced awkward array which makes it impossible to modify it. 

To solve this, it is necessary to perform a shallow copy of the awkward array on view creation (which is a cheap operation). I added this behavior here: 
https://github.com/scverse/anndata/pull/1070/files#diff-af8b8ed3c983f1ed77b81030eb9d80df8136e0640324f180354e6e0d3b49e7e7R139-R159

Maybe this could also be useful for pandas data frames with copy-on-write behavior in the future?

ivirshup commented:
I think caching the "view"s could be a good idea. I may need to think for a bit on how this effects my current thoughts on how to fix the garbage collection problem:

* https://github.com/scverse/anndata/pull/1119

Currently, this won't work with the solution proposed there, since the AlignedMappings are created on the fly to prevent circular references from being formed.
